src/dataset-operator/pkg/controller/datasetinternal/datasetinternal_controller.go

```diff
@@ -129,13 +131,65 @@ func (r *ReconcileDatasetInternal) Reconcile(request reconcile.Request) (reconci
 			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
 			// Return and don't requeue
 			reqLogger.Info("Dataset is not found")
-			err = nil
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
+	if instance.Spec.Local != nil {
+		datasetType := instance.Spec.Local["type"]
+		if (datasetType == "COS") {
+			if !contains(instance.GetFinalizers(), datasetFinalizer) {
+				if err := r.addFinalizer(reqLogger, instance); err != nil {
```

Comment:
Since ```addFinalizer``` updates the instance of the DatasetInternal (this update will trigger another reconcile) we should return always and not only when ```err != nil```.